File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
from collections import deque

# 상, 하, 좌, 우로 이동할 수 있는 좌표
dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


# BFS로 연결된 배추 군집을 탐색
def bfs(x, y, farm, visited, N, M):
    queue = deque()
    queue.append((x, y))
    visited[y][x] = True

    while queue:
        curr_x, curr_y = queue.popleft()

        # 상하좌우로 인접한 배추를 확인
        for i in range(4):
            nx = curr_x + dx[i]
            ny = curr_y + dy[i]

            # 범위 안에 있고, 배추가 있으며, 아직 방문하지 않은 경우
            if 0 <= nx < M and 0 <= ny < N and farm[ny][nx] == 1 and not visited[ny][nx]:
                visited[ny][nx] = True
                queue.append((nx, ny))


# 테스트 케이스 수 입력
T = int(input())

for _ in range(T):
    # M: 가로, N: 세로, K: 배추 위치의 개수
    M, N, K = map(int, input().split())

    # 배추밭 초기화 (0으로 초기화된 2D 배열)
    farm = [[0] * M for _ in range(N)]
    visited = [[False] * M for _ in range(N)]

    # 배추 위치 입력
    for _ in range(K):
        x, y = map(int, input().split())
        farm[y][x] = 1

    # 필요한 배추흰지렁이 마리 수 계산
    worm_count = 0

    for i in range(N):
        for j in range(M):
            # 배추가 있고 방문하지 않은 경우
            if farm[i][j] == 1 and not visited[i][j]:
                bfs(j, i, farm, visited, N, M)
                worm_count += 1

    # 결과 출력
    print(worm_count)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.util.*;
public class Main {

    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    public static void bfs(int x, int y, int[][] farm, boolean[][] visited, int m, int n) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{x, y});
        visited[x][y] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int curr_x = current[0];
            int curr_y = current[1];

            for (int i = 0; i < 4; i++) {
                int nx = curr_x + dx[i];
                int ny = curr_y + dy[i];

                if (nx >= 0 && nx < m && ny >= 0 && ny < n && farm[nx][ny] == 1 && !visited[nx][ny]) {
                    queue.add(new int[]{nx, ny});
                    visited[nx][ny] = true;
                }
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int T = sc.nextInt();

        for (int t = 0; t < T; t++) {
            int m = sc.nextInt();
            int n = sc.nextInt();
            int k = sc.nextInt();

            int[][] farm = new int[m][n];
            boolean[][] visited = new boolean[m][n];

            for (int i = 0; i < k; i++) {
                int x = sc.nextInt();
                int y = sc.nextInt();
                farm[x][y] = 1;
            }

            int wormCount = 0;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (farm[i][j] == 1 && !visited[i][j]) {
                        bfs(i, j, farm, visited, m, n);
                        wormCount++;
                    }
                }
            }
            System.out.println(wormCount);
        }
        sc.close();
    }
}
